rule map_paired_reads_with_bwa:
    conda:
        pipeline_path + "envs/bwa-samtools.yml"
    params:
        platform = "ILLUMINA"
    input:
        fastq1 = "samples/{sample}/reads/trimmed/R1_paired.fastq",
        fastq2 = "samples/{sample}/reads/trimmed/R2_paired.fastq",        
        ref = "references/{ref}/genome_fasta.fasta",
        ref_index = "references/{ref}/genome_fasta.fasta.bwt",
    output:
        temp("samples/{sample}/mapping/bwa/{ref}.bam"),
    log:
        logging_folder+"samples/{sample}/logs/mapping/bwa/{ref}/log.txt"
    shell:
        """
        if ls {output[0]}.tmp* 1> /dev/null 2>&1
        then
             rm {output[0]}.tmp*
        fi 
        (bwa mem -R '@RG\\tID:{wildcards.sample}\\tSM:{wildcards.sample}\\tPL:{params.platform}' {input[ref]} {input[fastq1]} {input[fastq2]} -v 1 | samtools sort -O BAM -o {output[0]}) 2> {log}
        """
        
rule map_single_reads_with_bwa:
    conda:
        pipeline_path + "envs/bwa-samtools.yml"
    params:
        platform = "ILLUMINA"
    input:
        fastq = "samples/{sample}/reads/trimmed/single.fastq",
        ref = "references/{ref}/genome_fasta.fasta",
        ref_index = "references/{ref}/genome_fasta.fasta.bwt",
    output:
        temp("samples/{sample}/mapping/bwa/{ref}.bam"),
    log:
        logging_folder+"samples/{sample}/logs/mapping/bwa/{ref}/log.txt"
    shell:
        """
        if ls {output[0]}.tmp* 1> /dev/null 2>&1
        then
             rm {output[0]}.tmp*
        fi 
        (bwa mem -R '@RG\\tID:{wildcards.sample}\\tSM:{wildcards.sample}\\tPL:{params.platform}' {input[ref]} {input[fastq]} -v 1 | samtools sort -O BAM -o {output[0]}) 2> {log}
        """


rule map_paired_reads_with_bwa_for_direct_sequencing:
    conda:
        pipeline_path + "envs/bwa-samtools.yml"
    params:
        platform = "ILLUMINA"
    input:
        fastq1 = "samples/{sample}/reads/trimmed/R1_paired.fastq",
        fastq2 = "samples/{sample}/reads/trimmed/R2_paired.fastq",        
        ref = "references/{ref}/genome_fasta.fasta",
        ref_index = "references/{ref}/genome_fasta.fasta.bwt",
    output:
        temp("samples/{sample}/mapping/bwa_stringent/{ref}.bam"),
    log:
        logging_folder+"samples/{sample}/mapping/bwa_stringent/{ref}/log.txt"
    shell:
        """
        if ls {output[0]}.tmp* 1> /dev/null 2>&1
        then
             rm {output[0]}.tmp*
        fi 
        (bwa mem -R '@RG\\tID:{wildcards.sample}\\tSM:{wildcards.sample}\\tPL:{params.platform}' {input[ref]} {input[fastq1]} {input[fastq2]} -v 1 -k 80 | samtools sort -O BAM -o {output[0]}) 2> {log}
        """
        
                
rule map_single_reads_with_bwa_for_direct_sequencing:
    conda:
        pipeline_path + "envs/bwa-samtools.yml"
    params:
        platform = "ILLUMINA"
    input:
        fastq = "samples/{sample}/reads/trimmed/single.fastq",
        ref = "references/{ref}/genome_fasta.fasta",
        ref_index = "references/{ref}/genome_fasta.fasta.bwt",
    output:
        temp("samples/{sample}/mapping/bwa_stringent/{ref}.bam"),
    log:
        logging_folder+"samples/{sample}/logs/mapping/bwa_stringent/{ref}/log.txt"
    shell:
        """
        if ls {output[0]}.tmp* 1> /dev/null 2>&1
        then
             rm {output[0]}.tmp*
        fi 
        (bwa mem -R '@RG\\tID:{wildcards.sample}\\tSM:{wildcards.sample}\\tPL:{params.platform}' {input[ref]} {input[fastq]} -v 1 -k 80 | samtools sort -O BAM -o {output[0]}) 2> {log}
        """

rule filter_reads_on_quality:
    conda:
        pipeline_path + "envs/bwa-samtools.yml"
    input:
        bam = "samples/{sample}/mapping/{mapping_method}/{ref}.bam",
        bam_bai = "samples/{sample}/mapping/{mapping_method}/{ref}.bam.bai",
        ref = "references/{ref}/genome_fasta.fasta",
        ref_index = "references/{ref}/genome_fasta.fasta.bwt",
    output:
        temp("samples/{sample}/mapping/{mapping_method}/{ref}_filtered.bam"),
    shell:
        """
        if ls {output[0]}.tmp* 1> /dev/null 2>&1
        then
            rm {output[0]}.tmp*
        fi 
        samtools view -q 60 -b -u -F 3844 -T {input[ref]} {input[bam]} | samtools sort -O bam -o {output[0]}
        """
        
rule remove_duplicates_from_mapping:
    conda:
        pipeline_path + "envs/gatk.yml"
    input:
        bam = "samples/{sample}/mapping/{mapping_method}/{ref}_filtered.bam",
        bam_bai = "samples/{sample}/mapping/{mapping_method}/{ref}_filtered.bam.bai",
    output:
        temp("samples/{sample}/mapping/{mapping_method}/{ref}_deduplicated_filtered.bam"),
    log:
         logging_folder + "samples/{sample}/mapping/{mapping_method}/{ref}/metrics_deduplicated.txt",
         logging_folder + "samples/{sample}/mapping/{mapping_method}/{ref}/markduplicates.txt",
    shell:
        """
        export _JAVA_OPTIONS=-Xmx3g && gatk-launch MarkDuplicates --REMOVE_DUPLICATES true -I {input[bam]} -O {output[0]} -M {log[0]} &> {log[1]}
        """

